glean-core/src/upload/result.rs

```diff
@@ -0,0 +1,66 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+/// Result values of attempted ping uploads encoded for FFI use.
+///
+/// In a perfect world this would live in `glean-ffi`,
+/// but because we also want to convert from pure integer values to a proper Rust enum
+/// using Rust's `From` and `Into` trait, we need to have it in this crate
+/// (The coherence rules don't allow to implement an external trait for an external type).
+///
+/// Due to restrictions of cbindgen they are re-defined in `glean-core/ffi/src/upload.rs`.
+///
+/// NOTE:
+/// THEY MUST BE THE SAME ACROSS BOTH FILES!
+pub mod ffi_upload_result {
+    /// A recoverable error.
+    pub const UPLOAD_RESULT_RECOVERABLE: u32 = 0x1;
+
+    /// An unrecoverable error.
+    pub const UPLOAD_RESULT_UNRECOVERABLE: u32 = 0x2;
+
+    /// A HTTP response code.
+    ///
+    /// The actual response code is encoded in the lower bits.
+    pub const UPLOAD_RESULT_HTTP_STATUS: u32 = 0x8000;
+}
+use ffi_upload_result::*;
+
+/// The result of an attempted ping upload.
+#[derive(Debug)]
+pub enum UploadResult {
+    /// A recoverable failure.
+    ///
+    /// During upload something went wrong,
+    /// e.g. the network connection failed.
+    /// The upload should be retried at a later time.
+    RecoverableFailure,
+
+    /// An unrecoverable upload failure.
+    ///
+    /// A possible cause might be a malformed URL.
+    UnrecoverableFailure,
+
+    /// A HTTP response code.
+    ///
+    /// This can still indicate an error, depending on the status code.
+    HttpStatus(u32),
+}
+
+impl From<u32> for UploadResult {
+    fn from(status: u32) -> Self {
+        match status {
+            status if (status & UPLOAD_RESULT_HTTP_STATUS) == UPLOAD_RESULT_HTTP_STATUS => {
+                // Extract the status code from the lower bits.
+                let http_status = status & !UPLOAD_RESULT_HTTP_STATUS;
```

Comment:
Also weird to see Rust deviate from the C family of bitwise operators and not go with the `~` for the logical NOT operation.

Reply:
As above, I wrote this with `~` first.
Then the compiler was helpful:

```
error: `~` cannot be used as a unary operator
 --> src/main.rs:3:17
  |
3 |     let x: u8 = ~0xf;
  |                 ^ help: use `!` to perform bitwise not
```	

`!` is indeed both bitwise and logical negation, and it's all driven by [a trait: `Not`](https://doc.rust-lang.org/std/ops/trait.Not.html).
But yes, it differs from C.

Reply:
Yup, I concur. A bit surprising, but I admit I like this more ... it feels consistent :)